### PR TITLE
OAuth: Enforce auto_assign_org_id setting when role mapping enabled using Generic OAuth

### LINKED ISSUE #22188: auto_assign_org_id is ignored for new generic OAuth users
**What happened**:
I log in via Generic OAuth with `auto_assign_org_id = 3` but the org I'm assigned is 1.

**What you expected to happen**:
I'd expect the value of `auto_assign_org_id` to be the organization I get upon first login.

**How to reproduce it (as minimally and precisely as possible)**:

Full config:

```
app_mode = production

[alerting]
enabled = false

[auth]
oauth_auto_login = true
signout_redirect_url = https://OMIT/auth/realms/osc/protocol/openid-connect/logout?redirect_uri=https%3A%2F%2FOMIT

[auth.generic_oauth]
allow_sign_up = true
api_url = https://OMIT/auth/realms/osc/protocol/openid-connect/userinfo
auth_url = https://OMIT/auth/realms/osc/protocol/openid-connect/auth
client_id = OMIT
client_secret = OMIT
enabled = true
name = Keycloak
role_attribute_path = contains(groups[*], 'sysstf') && 'Admin' || contains(groups[*], 'oscstaff') && 'Editor' || 'Viewer'
scopes = openid profile email groups
token_url = https://OMIT/auth/realms/osc/protocol/openid-connect/token

[auth.ldap]
allow_sign_up = true
config_file = /etc/grafana/ldap.toml
enabled = false

[database]
host = OMIT
name = grafana
password = OMIT
type = mysql
user = grafana

[security]
cookie_secure = true

[server]
domain = OMIT
http_port = 3000
root_url = https://OMIT

[users]
allow_sign_up = false
auto_assign_org = true
auto_assign_org_id = 3

[log]
level = debug
```

Login takes me to Keycloak then I get redirected to `https://grafana.DOMAIN/?orgId=1`

Logs:

```
Feb 13 16:34:58 grafana grafana-server: t=2020-02-13T16:34:58-0500 lvl=eror msg="Failed to look up user based on cookie" logger=context error="user token not found"
Feb 13 16:34:58 grafana grafana-server: t=2020-02-13T16:34:58-0500 lvl=info msg="Request Completed" logger=context userId=0 orgId=0 uname= method=GET path=/ status=302 remote_addr=LAPTOP_IP time_ms=1 size=29 referer=
Feb 13 16:34:58 grafana grafana-server: t=2020-02-13T16:34:58-0500 lvl=info msg="OAuth auto login enabled. Redirecting to /login/generic_oauth"
Feb 13 16:34:58 grafana grafana-server: t=2020-02-13T16:34:58-0500 lvl=info msg="Request Completed" logger=context userId=0 orgId=0 uname= method=GET path=/login status=307 remote_addr=LAPTOP_IP time_ms=1 size=56 referer=
Feb 13 16:34:58 grafana grafana-server: t=2020-02-13T16:34:58-0500 lvl=info msg="Request Completed" logger=context userId=0 orgId=0 uname= method=GET path=/login/generic_oauth status=302 remote_addr=LAPTOP_IP time_ms=0 size=325 referer=
Feb 13 16:34:58 grafana grafana-server: t=2020-02-13T16:34:58-0500 lvl=info msg="state check" logger=oauth queryState=8eefaea5bd356107c61d6ac37217bfced2daa2b902b0fd274ce63d597f825532 cookieState=8eefaea5bd356107c61d6ac37217bfced2daa2b902b0fd274ce63d597f825532
Feb 13 16:34:58 grafana grafana-server: t=2020-02-13T16:34:58-0500 lvl=dbug msg="OAuthLogin Got token" logger=oauth token="&{AccessToken:OMIT TokenType:Bearer RefreshToken:OMIT Expiry:2020-02-13 16:39:58.266131021 -0500 EST m=+1115.688870866 raw:map[access_token:OMIT expires_in:300 id_token:OMIT not-before-policy:0 refresh_expires_in:1800 refresh_token:OMIT scope:openid groups profile email session_state:cf6348b1-f4e3-4848-a395-04efc5840b7e token_type:bearer]}"
Feb 13 16:34:58 grafana grafana-server: t=2020-02-13T16:34:58-0500 lvl=dbug msg="Received id_token" logger=oauth.generic_oauth raw_json="{\"jti\":\"458b8e5a-49e3-4fc8-9870-7f8e91a490b1\",\"exp\":1581629998,\"nbf\":0,\"iat\":1581629698,\"iss\":\"https://idp.DOMAIN/auth/realms/osc\",\"aud\":\"grafana.DOMAIN\",\"sub\":\"f:OSC-LDAP-osc:tdockendorftest\",\"typ\":\"ID\",\"azp\":\"grafana.DOMAIN\",\"auth_time\":1581628420,\"session_state\":\"cf6348b1-f4e3-4848-a395-04efc5840b7e\",\"acr\":\"0\",\"email_verified\":false,\"name\":\"Trey Dockendorf Test\",\"groups\":[\"PZS0580\"],\"preferred_username\":\"tdockendorftest\",\"given_name\":\"Trey Dockendorf\",\"family_name\":\"Test\",\"email\":\"treydock@gmail.com\"}" data="Name: Trey Dockendorf Test, Displayname: , Login: , Username: , Email: treydock@gmail.com, Upn: , Attributes: map[]"
Feb 13 16:34:58 grafana grafana-server: t=2020-02-13T16:34:58-0500 lvl=dbug msg="HTTP GET https://idp.DOMAIN/auth/realms/osc/protocol/openid-connect/userinfo: 200 OK {\"sub\":\"f:OSC-LDAP-osc:tdockendorftest\",\"email_verified\":false,\"name\":\"Trey Dockendorf Test\",\"groups\":[\"PZS0580\"],\"preferred_username\":\"tdockendorftest\",\"given_name\":\"Trey Dockendorf\",\"family_name\":\"Test\",\"email\":\"treydock@gmail.com\"}"
Feb 13 16:34:58 grafana grafana-server: t=2020-02-13T16:34:58-0500 lvl=dbug msg="Received user info response" logger=oauth.generic_oauth raw_json="{\"sub\":\"f:OSC-LDAP-osc:tdockendorftest\",\"email_verified\":false,\"name\":\"Trey Dockendorf Test\",\"groups\":[\"PZS0580\"],\"preferred_username\":\"tdockendorftest\",\"given_name\":\"Trey Dockendorf\",\"family_name\":\"Test\",\"email\":\"treydock@gmail.com\"}" data="Name: Trey Dockendorf Test, Displayname: , Login: , Username: , Email: treydock@gmail.com, Upn: , Attributes: map[]"
Feb 13 16:34:58 grafana grafana-server: t=2020-02-13T16:34:58-0500 lvl=dbug msg="User info result" logger=oauth.generic_oauth result="&{Id: Name:Trey Dockendorf Test Email:treydock@gmail.com Login:treydock@gmail.com Company: Role:Viewer Groups:[]}"
Feb 13 16:34:58 grafana grafana-server: t=2020-02-13T16:34:58-0500 lvl=dbug msg="OAuthLogin got user info" logger=oauth userInfo="&{Id: Name:Trey Dockendorf Test Email:treydock@gmail.com Login:treydock@gmail.com Company: Role:Viewer Groups:[]}"
Feb 13 16:34:58 grafana grafana-server: t=2020-02-13T16:34:58-0500 lvl=dbug msg="user auth token created" logger=auth tokenId=56 userId=27 clientIP=LAPTOP_IP userAgent="Mozilla/5.0 (Macintosh; Intel Mac OS X 10.14; rv:72.0) Gecko/20100101 Firefox/72.0" authToken=769674a96dd0dd0ad213765e2e1c04518b873000ac553a189088e8d799520c3d
Feb 13 16:34:58 grafana grafana-server: t=2020-02-13T16:34:58-0500 lvl=info msg="Successful Login" logger=http.server User=treydock@gmail.com
Feb 13 16:34:58 grafana grafana-server: t=2020-02-13T16:34:58-0500 lvl=info msg="Request Completed" logger=context userId=0 orgId=0 uname= method=GET path=/login/generic_oauth status=302 remote_addr=LAPTOP_IP time_ms=249 size=24 referer=
Feb 13 16:34:58 grafana grafana-server: t=2020-02-13T16:34:58-0500 lvl=dbug msg="seen token" logger=auth tokenId=56 userId=27 clientIP=LAPTOP_IP userAgent="Mozilla/5.0 (Macintosh; Intel Mac OS X 10.14; rv:72.0) Gecko/20100101 Firefox/72.0" authToken=769674a96dd0dd0ad213765e2e1c04518b873000ac553a189088e8d799520c3d
Feb 13 16:34:58 grafana grafana-server: t=2020-02-13T16:34:58-0500 lvl=dbug msg="Updating last user_seen_at" logger=context userId=27 orgId=1 uname=treydock@gmail.com user_id=27
```

**Anything else we need to know?**:

The user is correctly assigned the "Viewer" role but the wrong org is getting assigned. This is a problem because we want our user community to only have access to ordId=3 which is dashboards approved for our user community. ordId=1 is for staff only.  A work around would be to change the folder permissions of all folders in orgId=1 to remove the Viewer but that seems like it should be unnecessary.

**Environment**:
- Grafana version: 6.6.1
- Data source type & version: Prometheus 2.15.2
- OS Grafana is installed on: RHEL 7.7
- User OS & Browser: Mac and Firefox
- Grafana plugins: Image render
- Others:

### PR DESCRIPTION
**Pull request by aknuds1**
**What this PR does / why we need it**:
Use auto_assign_org_id setting, if set, to determine organization an OAuth user will be member of.

**Which issue(s) this PR fixes**:
Fixes #22188.

### CODE REVIEW

**Review by papagian — APPROVED** [review]
I have tried creating user by logging to Okta by setting `auto_assign_org_id` to:
-  an existing orgId (other than 1)
- non-existing orgId
and it in the first case I got assigned to the correct organisation and in the latter I got an error `Could not create user: organization id 3 does not exist`

I tried also the same in master and I was faulty assigned to the orgId = 1.

Great work fixing this!

**Comment by marefr** [comment]
@papagian what do you mean by first case and the latter? master vs these changes? What you've described sounds like a bug to me?

**Comment by papagian** [comment]
> @papagian what do you mean by first case and the latter? master vs these changes? What you've described sounds like a bug to me?

Sorry, I rephrase it to be more clear:

I have tried creating user by logging to Okta by setting `auto_assign_org_id` to:
- an existing orgId (orgId=4); I got assigned to the correct organisation with orgId=4.
- non-existing orgId; I got an error `Could not create user: organization id 3 does not exist`

I tried also the same in master and I was faulty assigned to the orgId = 1.

**Comment by aknuds1** [comment]
> I have tried creating user by logging to Okta by setting `auto_assign_org_id` to:
> 
>     * an existing orgId (other than 1)
> 
>     * non-existing orgId
>       and it in the first case I got assigned to the correct organisation and in the latter I got an error `Could not create user: organization id 3 does not exist`
> 
> 
> I tried also the same in master and I was faulty assigned to the orgId = 1.
> 
> Great work fixing this!

Thanks for testing the PR @papagian :) I haven't got that far yet; not sure how to drive OAuth/Grafana, and got busy fixing a bug in my pipeline.

File: pkg/services/login/login.go [hunk 1]
```diff
@@ -112,16 +108,14 @@ func (ls *LoginService) UpsertUser(cmd *models.UpsertUserCommand) error {
 		}
 	}
 
-	err = ls.Bus.Dispatch(&models.SyncTeamsCommand{
+	if err := ls.Bus.Dispatch(&models.SyncTeamsCommand{
```
Comment (marefr) [inline pkg/services/login/login.go]: Nit. Not sure this reads better than before. I suggest moving the `if and condition` below `ls.Bus.Dispatch(&models.SyncTeamsCommand{...`

**Review by marefr — APPROVED** [review]
LGTM 

The only thing left is whether `auto_assign_org=false` should populate an org role at all, because that would fail login if org id 1 doesn't exist and/or maybe user have configured it this way so that user should get their own personal organization. But lets wait with this since it should be an edge case using OAuth role mapping and `auto_assign_org=false`.

Added a minor style comment/nit.

File: pkg/services/login/login.go (hunk 1 above)
Comment (aknuds1) [reply pkg/services/login/login.go]: @marefr OK I can make the change; the idea was to scope `err` locally to the `if` statement, it helps with avoiding usage of dangling `err`s.
